app/src/main/java/com/android/partagix/model/Database.kt

```diff
@@ -389,6 +392,117 @@ class Database(database: FirebaseFirestore = Firebase.firestore) {
     onSuccess(user)
   }
 
+  /**
+   * Retrieve all comments that a user has received, both as a lender than as an borrower
+   *
+   * @param userId the user's id
+   * @param onSuccess the function to return a List containing pairs (comment's author name,
+   *   comment)
+   */
+  fun getComments(userId: String, onSuccess: (List<Pair<String, String>>) -> Unit) {
+    val ret = mutableListOf<Pair<String, String>>()
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLender == userId // if the user is the lender,
+            &&
+            loan.reviewLender.toDouble() != 0.0 // that has been reviewed,
+            &&
+            loan.commentLender != "") { // and received a comment
+
+          getUser(loan.idBorrower, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentLender))
+          }
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idBorrower == userId // if the user is the borrower,
+            &&
+            loan.reviewBorrower.toDouble() != 0.0 // that has been reviewed
+            &&
+            loan.commentBorrower != "") { // and received a comment
+
+          getUser(loan.idLender, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentBorrower))
+          }
+        }
+      }
+    }
+
+    onSuccess(ret)
+  }
+
+  /**
+   * Retrieve all ranks that a user has received, both as a lender than as an borrower, compute the
+   * average rank, and store it in the user's rank
+   *
+   * @param idUser the user's id
+   */
+  fun newAverageRank(idUser: String) {
+
+    var rankSum = 0.0
+    var rankCount = 0L
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLender == idUser // if the user is the lender,
+            &&
+            loan.reviewLender.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewLender.toDouble()
+          rankCount++
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idBorrower == idUser // if the user is the borrower,
+            &&
+            loan.reviewBorrower.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewBorrower.toDouble()
+          rankCount++
+        }
+      }
+      val averageRank = rankSum / rankCount.toDouble()
+      users.document(idUser).update("rank", averageRank.toString())
```

Comment:
maybe we could store the rounded average rank and not the result of the division but this is just a suggestion and is discutable

Reply:
This is already handled by the UI part, hence we'll keep the result as the double.

Reply:
it would be an operation done one time instead of every time the ui loads. It's also ok like that.